doc/design.md

```diff
@@ -0,0 +1,258 @@
+
+Because the Python UDF cannot return values to the calling Lua Script, before 
+their query ends, we have to call for each iteration the UDF again.
+- This means the UDF has to store and load its state each time it is getting called. 
+- Furthermore, the interface to the user code needs to be suitable for this form of execution. 
+- The user code can't wait actively for the result of a query. The framework will execute it while the UDF is running anymore.
+    - The user code has to get called again, when the SQL queries are ready. 
+    - The user code is a kind of callback, however a call back with state.
+    - To this execution, fits best the model of an event handler.
+    - The Lua Script is then our event loop.
+
+## System Design and Architecture
+ 
+The designed event-driven framework is consists of 3 
+main components: (1) Even Loop that handles only the state transitions, (2) 
+Event Handler Framework that defines a state machine and provide a framework to 
+user code, (3) Event Handler, where user implements algorithm. The Event Loop 
+component is proposed to be implemented in Lua, because Lua script  is the only 
+way to run the dynamic SQL queries in the same transactions. On the other hand, 
+Event Handler is implemented in Python  cript, since Python simplifies the 
+development of data analysis methods by offering a wide variety of data 
+processing tools.
+
+
+## Event Loop
+Event Loop  processes only the state transitions by executing queries returned 
+by the Event Handler. This loop provides the new state transition by calling 
+the Event Handler until the Event Handler returns the stop signal with a result 
+or an error.
+
+### Initiating Event Loop
+`dsn~initiating-event-loop~1`
+
+The Event Loop is responsible for getting the parameters of the Event Handler 
+that will be called in the loop and forward them to it.
+
+Covers:
+
+* `req~intiating-algorithm~1`
+
+Needs: impl, utest, itest
+
+### Iterating over Loop
+`dsn~handling-states~1`
+
+The  output of the called Event Handler query can include SQL queries to 
+complete its computation. In that case Event Loop run them on each iteration, 
+until the Event Handler returns the status as complete.
+
+Covers:
+
+* `req~iterating-over-loop~1`
+
+Needs: impl, utest, itest
+
+
+
+### Returning JSON Result
+`dsn~returning-json-result~1`
+
+The Event Loop returns result in json format.
+
+Covers:
+
+* `req~returning-result~1`
+
+Needs: impl, utest, itest
+
+
+
+## Event Handler
+The Event Handler is a framework whose logic is implemented by users. It reacts 
+to the query results of  the Event Loop and generates SQL queries of the next 
+state. It stores the states in BucketFS.
+
+
+### Python UDF Framework 
+`dsn~python-udf-framework~1`
+
+An Event Handler designed as Python UDF  allows users to have an easy-to-use and 
+better tooling framework to build their algorithm on top it. The user can develop 
```

Comment:
```suggestion
An Event Handler designed as Python UDF instead of a Lua script allows users to have an easy-to-use and 
better tooling framework to build their algorithm on top of. The user can develop 
```